nrf-softdevice/src/util/portal.rs

```diff
@@ -23,137 +26,144 @@ enum State<T> {
 unsafe impl<T> Send for Portal<T> {}
 unsafe impl<T> Sync for Portal<T> {}
 
+#[cfg(not(feature = "usable-from-interrupts"))]
 fn assert_thread_mode() {
     assert!(
         cortex_m::peripheral::SCB::vect_active() == cortex_m::peripheral::scb::VectActive::ThreadMode,
-        "portals are not usable from interrupts"
+        "portals are not usable from interrupts. Activate the 'usable-from-interrupts' feature to allow this."
     );
 }
 
 impl<T> Portal<T> {
     pub const fn new() -> Self {
         Self {
-            state: UnsafeCell::new(State::None),
+            state: Mutex::new(RefCell::new(State::None)),
         }
     }
 
     pub fn call(&self, val: T) -> bool {
+        #[cfg(not(feature = "usable-from-interrupts"))]
         assert_thread_mode();
-
-        // safety: this runs from thread mode
-        unsafe {
-            match *self.state.get() {
-                State::None => false,
-                State::Done => false,
-                State::Running => panic!("Portall::call() called reentrantly"),
-                State::Waiting(func) => {
-                    (*func)(val);
-                    true
-                }
-            }
+        let maybe_func = self.state.lock(|state| match *state.borrow() {
+            State::None => None,
+            State::Done => None,
+            State::Running => panic!("Portal::call() called reentrantly"),
+            State::Waiting(func) => Some(func),
+        });
+
+        // re-entrant calling possible here. Acceptable because Portal::call() panics.
+
+        if let Some(ptr) = maybe_func {
+            // Safety: This is transmuted from a FnMut, and therefore valid
+            unsafe { (*ptr)(val) };
         }
+        true
     }
 
-    pub fn wait_once<'a, R, F>(&'a self, mut func: F) -> impl Future<Output = R> + 'a
+    pub async fn wait_once<'a, R, F>(&'a self, mut func: F) -> R
     where
         F: FnMut(T) -> R + 'a,
     {
+        #[cfg(not(feature = "usable-from-interrupts"))]
         assert_thread_mode();
+        let signal = Signal::<CriticalSectionRawMutex, _>::new();
+        let mut result: MaybeUninit<R> = MaybeUninit::uninit();
 
-        async move {
-            let signal = Signal::<CriticalSectionRawMutex, _>::new();
-            let mut result: MaybeUninit<R> = MaybeUninit::uninit();
-            let mut call_func = |val: T| unsafe {
-                let state = &mut *self.state.get();
-
+        let mut call_func = |val: T| unsafe {
+            self.state.lock(|state| {
+                let mut state = state.borrow_mut();
                 // Set state to Running while running the function to avoid reentrancy.
```

Comment:
now that there's a RefCell, it's not really possible for this to be called reentrantly, because it'll be caught by `.borrow_mut()`. Perhaps it's possible to simplify the state?

Reply:
Might be possible by passing the mutable reference to the callback directly

Reply:
That should guarantee that only the one borrow exists until the closure is called properly, excluding everything that comes after the execution (which would really just be returning true, which should be completely fine to do concurrently with a new method start).
I'll try to rewrite this to work that way

Reply:
I think the whole `enum State` can be replaced with a `Option<*mut dyn FnMut(T)>`.
- The `Done` state is not doing much, it's equivalent to `None`.
- The `Running` state is no longer needed, we're protected by the RefCell.

Reply:
New version fulfills this, and should be faster overall.
Also takes out 60 lines or so, which should be helpful for keeping this maintainable